keps/sig-apps/4212-improve-node-maintenance/README.md

```diff
@@ -0,0 +1,1065 @@
+
+### Disruption Controller: Eviction Protection and Observability
+
+With the `.spec.MinAvailable` and `.spec.MaxUnavailable` options, PDB could allow evicting the surge
+pods because `.status.CurrentHealthy` could become higher than `.status.desiredHealthy`. This could
+disrupt ongoing migration of an application. To support mission-critical applications, we would add
+a `.spec.maxSurge` field. This would cause the disruption controller to count the
+`.status.desiredHealthy` and `.status.DisruptionsAllowed` with max surge in mind. This field could
+serve two purposes:
+
+1. If set to true, it could change the counting of status fields based on the pod workload
+   maxSurge (deployments) or based on the Surge API (see below)
+2. If set to a number (int or percentage), it would override the detected maxSurge and use only the
+   maxSurge value in the PDB object.
+
+
+```go
+type PodDisruptionBudgetSpec struct {
+    ...
+    // if true, prevents disruption of all detected surge pods
+	// if integer or percentage, prevents disruption of the defined amount of surge pods
+    MaxSurge int32|string|bool // can we do BoolOrIntOrString ?
+    ...
+}
+
+type PodDisruptionBudgetStatus struct {
+	...
+	// total number of pods that are expected to surge during a node maintenance
+    ExpectedSurgePods int32
+	...
+}
+```
+
+
+The disruption controller should check whether a workload guarded by a PDB has a pod surge capability.
+If it does, it should report `.status.expectedSurgePods` to allow cluster admins to assess which PDBs
+are blocking cluster upgrade. If the `.status.disruptionsAllowed` equals to 0, because for example
+`.spec.maxUnavailable` is equal to 0, the cluster admin can assess that the PDB will not block
+cluster upgrade if he or she observes `.status.expectedSurgePods` greater than 0.
+
+
+### Surge API (Optional)
+
+Any CustomResourceDefinition or built-in type could implement a surge API. This could be a
+subresource similar to a scale API, that would serve for advertising the surge support. If a
+certain workload supports a pod surge, then one of `.status.availableSurgeReplicas` or
+`.status.currentSurgeReplicas` values will be positive.
+
+```go
+
+type Surge struct {
+	...
+	Status SurgeStatus
+}
+
+
+type SurgeStatus struct {
+	// number of replicas that can surge during an application rollout or node maintenance
+    SurgeReplicas int32
+	// current number of surge replicas that are running above the normal amount of replicas
+	CurrentSurgeReplicas int32
+}
+
+```
+
+This API would be then used by kubectl and disruption controller to reason about the capabilities
+of an application. It would be similar to how we use the scale API in podControllerFinders in
+disruption controller.
+
+### Node Drain Process: Before and After
+
+The following diagrams describe how the node drain process will change in respective to each
+component.
+
+Current state of node drain:
+
+![drain-before](https://kubernetes.io/images/docs/kubectl_drain.svg)
```

Comment:
I recommend copying that diagram into _this_ repo.

Reply:
updated with a better link, see discussion: https://github.com/kubernetes/enhancements/pull/4213#discussion_r1333444863